website/addons/googledrive/model.py

```diff
@@ -2,235 +2,189 @@
-        self.refresh_access_token(force=force_refresh)
-        return self.access_token
+    client_id = drive_settings.CLIENT_ID
+    client_secret = drive_settings.CLIENT_SECRET
 
-    def refresh_access_token(self, force=False):
-        if self._needs_refresh() or force:
-            client = GoogleAuthClient()
-            token = client.refresh(self.access_token, self.refresh_token)
+    auth_url_base = "https://accounts.google.com/o/oauth2/auth?access_type=offline&approval_prompt=force"
+    callback_url = 'https://www.googleapis.com/oauth2/v3/token'
 
-            self.access_token = token['access_token']
-            self.refresh_token = token['refresh_token']
-            self.expires_at = datetime.utcfromtimestamp(token['expires_at'])
-            self.save()
+    default_scopes = drive_settings.OAUTH_SCOPE
+    _auth_client = GoogleAuthClient()
+    _drive_client = GoogleDriveClient()
 
-    def revoke_access_token(self):
-        # if there is only one osf user linked to this google drive user oauth, revoke the token,
-        # otherwise, disconnect the osf user from the googledriveoauthsettings
-        if len(self.googledriveusersettings__accessed) <= 1:
-            client = GoogleAuthClient()
-            try:
-                client.revoke(self.access_token)
-            except:
-                # no need to fail, revoke is opportunistic
-                pass
+    def handle_callback(self, response):
+        client = self._auth_client
+        info = client.userinfo(response['access_token'])
+        return {
+            'provider_id': info['sub'],
+            'display_name': info['name'],
+            'profile_url': info.get('profile', None)
+        }
 
-            # remove the object as its the last instance.
-            GoogleDriveOAuthSettings.remove_one(self)
+    def _refresh_token(self, access_token, refresh_token):
+        """ Handles the actual request to refresh tokens
 
-    def _needs_refresh(self):
-        if self.expires_at is None:
+        :param str access_token: Access token (oauth key) associated with this account
+        :param str refresh_token: Refresh token used to request a new access token
+        :return dict token: New set of tokens
+        """
+        client = self._auth_client
+        if refresh_token:
+            token = client.refresh(access_token, refresh_token)
+            return token
+        else:
             return False
-        return (self.expires_at - datetime.utcnow()).total_seconds() < drive_settings.REFRESH_TIME
 
+    def fetch_access_token(self, force_refresh=False):
+        self.refresh_access_token(force=force_refresh)
+        return self.account.oauth_key
 
-class GoogleDriveUserSettings(AddonUserSettingsBase):
-    """Stores user-specific information, including the Oauth access
-    token.
-    """
-    oauth_settings = fields.ForeignField(
-        'googledriveoauthsettings', backref='accessed'
-    )
+    def refresh_access_token(self, force=False):
```

Comment:
Not every OAuth implementation uses refresh tokens, but it's common enough that refresh logic should not go in a single sub-class. It should be available as part of the parent architecture (perhaps a `RefreshTokenProvider` subclass or such)

If you think that's a worthwhile future change, can you create (and link me to) the standalone JIRA ticket for that refactor? It won't block this PR but would be nice to have in the future.